checkers/typeDefFirst_checker.go

```diff
@@ -0,0 +1,75 @@
+package checkers
+
+import (
+	"go/ast"
+	"go/token"
+	"strings"
+
+	"github.com/go-critic/go-critic/checkers/internal/astwalk"
+	"github.com/go-critic/go-critic/framework/linter"
+)
+
+func init() {
+	var info linter.CheckerInfo
+	info.Name = "typeDefFirst"
+	info.Tags = []string{"style", "experimental"}
+	info.Summary = "File-scoped checker, that requires type definition before its method definitions"
+	info.Before = `
+func (r rec) Method() {}
+type rec struct{}
+`
+	info.After = `
+type rec struct{}
+func (r rec) Method() {}
+`
+	collection.AddChecker(&info, func(ctx *linter.CheckerContext) linter.FileWalker {
+		return &typeDefFirstChecker{
+			ctx: ctx,
+		}
+	})
+}
+
+type typeDefFirstChecker struct {
+	astwalk.WalkHandler
+	ctx *linter.CheckerContext
+}
+
+func (t *typeDefFirstChecker) WalkFile(f *ast.File) {
+	typeUsageMap := make(map[string]bool)
+
+	if f.Decls == nil {
+		return
+	}
+
+	for _, declaration := range f.Decls {
+		switch decl := declaration.(type) {
+		case *ast.FuncDecl:
+			if decl.Recv != nil {
+				receiver := decl.Recv.List[0]
```

Comment:
should I check the array for not nilness? as far as I understood, in this case (where its receiver) there are always should be some fields, but the safer the better?

Reply:
you can add a simple `func` in golden files if it fails - better to add :)

Reply:
It does not fail, because it goes after `decl.Recv != nil` where Recv is recivier